### ISSUE TITLE
Test the hypothesis that calls Cgo significantly affects the chat performance

### ISSUE THREAD

**Issue by JekaMas** [status-im/status-go#579]
# Problem

Ethereum packages: `ethereum/go-ethereum/crypto/ecies`, `ethereum/go-ethereum/crypto/sha3`, `ethereum/go-ethereum/crypto/secp256k1` take from 40 up to 55% CPU time while message sending-receiving (tested on localhost with 2 Status nodes, without mail server, on ropsten network). They use cgo so each call can causes fork and stack copy.

# Implementation

Check the hypothesis on user cases from idea https://github.com/status-im/ideas/issues/71

# Acceptance Criteria

Defenite answer does this hypothesis take place on our user cases.

# Notes

# Future Steps

If answer is 'yes' make a new issue to resolve bottleneck on Ethereum side.

**Comment by feuGeneA** [comment]
While I haven't tested this hypothesis specifically against the use cases in [status-im/ideas#71](https://github.com/status-im/ideas/issues/71), I have managed to get some measurements about the overhead of CGo in general.

In short, the CGo overhead seems negligible in comparison to the amount of time spent within the underlying C routines.  All of the gory, supporting details are below.

(As an aside, it seems as though only one of the `go-ethereum` packages mentioned in the OP -- namely, `ethereum/go-ethereum/crypto/secp256k1` -- is actually using CGo.  I'm not sure how the other two got onto that list.)

While it would seem intuitive to use the built-in Go CPU profiling, it does not work well for this case.  For one thing, despite some mentions on the web that it "will be possible in Go v1.8," the built-in CPU profiling does *not* capture information about different C calls made from a `cgocall`, so from the `pprof` perspective there is no way to differentiate CGo overhead from actual C library usage:
![cgocall](https://user-images.githubusercontent.com/7883777/35742798-9f418b5c-0809-11e8-9f9c-17b3d53b7c18.png)
The other problem, with using Go's built-in CPU profiling for this case, is the matter of sample collection.  While doing CPU profiling, Go takes samples of the call stack 100 times per second, and it uses those samples as the basis for its performance reporting.  When using existing messaging tests (specifically, `status-go/e2e/whisper`) there is much down time, when the execution is presumably I/O bound, as you can see in the `go tool trace` output below.  Also, there are lots of other things going on in this test, activity which does not use such CGo-wrapped crypto routines.  Consequently, it is hard to capture a `pprof` output that contains the crypto calls of interest.  If I run it 5 times, only once or twice will the `cgocall` and `secp256k` stuff be visible.
![whispertrace](https://user-images.githubusercontent.com/7883777/35736793-50b5c0b4-07f7-11e8-9bc0-441a8c36bafd.png)
While Go's built-in CPU profiling doesn't shed any light on C routine usage, the Linux `perf` tool does, as it samples the call stack much more frequently, at 1000 times per second [(ref)](https://perf.wiki.kernel.org/index.php/Tutorial#Period_and_rate).  You can see exactly what `libsecp256k` routines are being used, and how much of the time is being spent in them:
```
gene@precision5510:/home/gene/go/src/github.com/status-im/status-go$ go test -c ./e2e/whisper/...
gene@precision5510:~/go/src/github.com/status-im/status-go$ perf record ./whisper.test -network=
...
gene@precision5510:/home/gene/go/src/github.com/status-im/status-go$ perf report -f --stdio --symbol-filter=secp
# Total Lost Samples: 0
#
# Samples: 7K of event 'cycles:pp'
# Event count (approx.): 4360258211
#
# Overhead  Command       Shared Object  Symbol                                 
# ........  ............  .............  .......................................
#
     0.78%  whisper.test  whisper.test   [.] secp256k1_fe_mul_inner             
     0.20%  whisper.test  whisper.test   [.] secp256k1_fe_sqr_inner             
     0.06%  whisper.test  whisper.test   [.] secp256k1_fe_normalize             
     0.06%  whisper.test  whisper.test   [.] secp256k1_ge_set_gej_zinv          
     0.05%  whisper.test  whisper.test   [.] secp256k1_fe_normalize_weak        
     0.03%  whisper.test  whisper.test   [.] secp256k1_scalar_reduce_512        
     0.02%  whisper.test  whisper.test   [.] secp256k1_gej_double_var           
     0.02%  whisper.test  whisper.test   [.] secp256k1_fe_inv                   
     0.02%  whisper.test  whisper.test   [.] secp256k1_gej_add_ge               
     0.02%  whisper.test  whisper.test   [.] secp256k1_gej_add_ge_var           
     0.01%  whisper.test  whisper.test   [.] secp256k1_fe_normalizes_to_zero_var
     0.01%  whisper.test  whisper.test   [.] secp256k1_gej_add_var.constprop.17 
     0.01%  whisper.test  whisper.test   [.] secp256k1_ecmult_wnaf.constprop.15 
     0.01%  whisper.test  whisper.test   [.] secp256k1_scalar_mul_512           
```
But the `perf` output from `go test e2e/whisper/...` doesn't answer the CGo overhead question, so I wrote the following benchmark of `go-ethereum/crypto.ToECDSA()`, which under the covers utilizes `secp256k`.
```go
func BenchmarkECDSA(b *testing.B) {
	var privateKey *ecdsa.PrivateKey
	privateKey, _ = crypto.GenerateKey()

	var binDump []byte
	binDump = crypto.FromECDSA(privateKey)

	for n := 0; n < b.N; n++ {
		crypto.ToECDSA(binDump)
	}
}
```
As the following `pprof` output shows, this benchmark achieves the the intensive CGo usage that we need to measure.
![profile001](https://user-images.githubusercontent.com/7883777/35745874-c3ce335e-0812-11e8-9ecb-1f261111c0ce.png)
And here is the `perf report` ouptut for the same test executable:
```
gene@precision5510:~/go/src/github.com/status-im/status-go/e2e/whisper$ perf report --symbol-filter=secp --stdio
# Total Lost Samples: 0
#
# Samples: 6K of event 'cycles:pp'
# Event count (approx.): 5111381108
#
# Overhead  Command       Shared Object  Symbol                                                                                         
# ........  ............  .............  .............................................................................................................................
#
    34.86%  whisper.test  whisper.test   [.] secp256k1_fe_mul_inner                                                                                                   
    28.77%  whisper.test  whisper.test   [.] secp256k1_fe_sqr_inner                                                                                                   
     9.94%  whisper.test  whisper.test   [.] secp256k1_gej_double_var                                                                                                 
     4.97%  whisper.test  whisper.test   [.] secp256k1_fe_inv                                                                                                         
     4.64%  whisper.test  whisper.test   [.] secp256k1_pubkey_scalar_mul                                                                                              
     3.97%  whisper.test  whisper.test   [.] secp256k1_gej_add_ge                                                                                                     
     1.88%  whisper.test  whisper.test   [.] secp256k1_fe_normalize_weak                                                                                              
     0.72%  whisper.test  whisper.test   [.] secp256k1_fe_set_b32                                                                                                     
     0.65%  whisper.test  whisper.test   [.] secp256k1_fe_normalizes_to_zero                                                                                          
     0.57%  whisper.test  whisper.test   [.] secp256k1_fe_get_b32                                                                                                     
     0.41%  whisper.test  whisper.test   [.] secp256k1_gej_add_ge_var                                                                                                 
     0.12%  whisper.test  whisper.test   [.] secp256k1_fe_normalize                                                                                                   
     0.12%  whisper.test  whisper.test   [.] secp256k1_ge_set_gej_zinv                                                                                                
     0.11%  whisper.test  whisper.test   [.] secp256k1_scalar_shr_int.constprop.20                                                                                    
     0.06%  whisper.test  whisper.test   [.] secp256k1_ge_set_gej                                                                                                     
     0.06%  whisper.test  whisper.test   [.] secp256k1_fe_normalizes_to_zero_var                                                                                      
     0.05%  whisper.test  whisper.test   [.] secp256k1_gej_set_ge                                                                                                     
     0.03%  whisper.test  whisper.test   [.] secp256k1_fe_to_storage                                                                                                  
     0.03%  whisper.test  whisper.test   [.] secp256k1_context_create                                                                                                 
     0.02%  whisper.test  whisper.test   [.] github.com/status-im/status-go/vendor/github.com/ethereum/go-ethereum/crypto/secp256k1.(*BitCurve).ScalarMult.func1      
     0.02%  whisper.test  whisper.test   [.] secp256k1_scalar_negate                                                                                                  
     0.01%  whisper.test  whisper.test   [.] github.com/status-im/status-go/vendor/github.com/ethereum/go-ethereum/crypto/secp256k1._Cfunc_secp256k1_pubkey_scalar_mul
     0.01%  whisper.test  whisper.test   [.] secp256k1_ge_globalz_set_table_gej.constprop.18                                                                          
```
(Note: Adding all those percentages yields 92.02%.)  Note that the symbol `github.com/status-im/status-go/vendor/github.com/ethereum/go-ethereum/crypto/secp256k1._Cfunc_secp256k1_pubkey_scalar_mul` appears in both the `pprof` and `perf` outputs.  Presumably this symbol represents the "CGo overhead," and as you can see above, the amount of time preparing for the C call is negligible compared to the amount of time spent in the C routines themselves.

Also interesting is the output from running the benchmark:
```
$ ./whisper.test -test.bench=. -test.run=XXX
goos: linux
goarch: amd64
pkg: github.com/status-im/status-go/e2e/whisper
BenchmarkECDSA-8   	   10000	    146988 ns/op
PASS
```
Despite the commonly held notion that repeated calls from Go to a C routine can incur large penalties due to repeated forking and call stack copying, it seems to be a non-issue here, as apparently a single call to `crypto.ToECDSA()` (that is, a single iteration of the benchmark's loop over `b.N`) completes in just 0.146988 ms.

**Comment by JekaMas** [comment]
Awesome! Your C expiriense is really great.
Did you use whisper tests without any modifications?

**Comment by feuGeneA** [comment]
Thanks!  Yes, I used the e2e/whisper/... tests without any modifications.

**Comment by JekaMas** [comment]
@feuGeneA Could it be that the amount of messages from tests are not statistically significant? As far as i can see each test sends just 1-3 message from node to node.
Just to be sure, could you make a new benchmark with greater message rps?
You can take the 4th case from https://github.com/status-im/ideas/issues/71 :

- run 2 nodes (mail server isn't necessary)
- check that nodes are connected with a network (`backend.NodeManager().PeerCount() > 0)
- create a symKey (for example)
- make a new topic
- add the the key on each node
- make a message filter by the topic on each node
- send messages between the nodes with frequence 1 message from each node per second
- check does each filter obtains new messages on each step of message exchange
- code all steps as golang benchmark
- run test on Rinkeby network with upstream
- check the results

As the example you could take `TestRequestMessagesInGroupChat`, all methods to send, get messages, create filters, nodes are in `e2e/whisper/whisper_mailbox_test.go`.

**Comment by feuGeneA** [comment]
In short:

- CGo overhead still looks negligible.
- I turned up a couple of problems that may warrant further investigation (please weigh in):
  - `CRIT [02-09|02:20:06] Failed to store last header's hash       err="leveldb: closed"`, easily reproduced with a long-running (10m) benchmark.
  - `fatal error: concurrent map read and map write`, raised by a `go-ethereum/les` func, which I only encountered once but which may be reproducible by contriving some contention in that `les` code.
- I'm not sure if we should commit my benchmark code.  Please advise.

Details below.

---
You can see the benchmark I used [here](https://github.com/status-im/status-go/blob/tests/%23579-benchmark-whisper-for-cgo/t/e2e/whisper/whisper_mailbox_test.go#L363).  Essentially, it creates a 3-party group chat, and approximately once per second it has party A send a message and parties B and C receive it.

(Please let me know if you think this benchmark is worth merging into the codebase, or if we should treat it as a one-off.  If it's to be merged in, it'll require a bit of refactoring, since I had to comment out all of the `testify/suite/Suite.Require()` calls (and similar verifications) in `WhisperMailboxSuite`s `startBackend()` and `startMailboxBackend()` methods, since a benchmark doesn't receive a `testing.T` object that can be passed through to that existing code.)

I wanted to run the benchmark for 10 minutes, but this error precluded me from doing that:
```
gene@precision5510:~/go/src/github.com/status-im/status-go/e2e/whisper$ ./whisper.test -test.bench=. -test.run=XXX -test.benchtime=600s -network=4 -test.cpuprofile=whisper.pprof
goos: linux
goarch: amd64
pkg: github.com/status-im/status-go/e2e/whisper
BenchmarkRequestMessagesInGroupChat-8   	CRIT [02-09|02:20:06] Failed to store last header's hash       err="leveldb: closed"
```
I found that I could run the test for 400 seconds (6.67 minutes), so I went with that for this benchmarking, but should we separately investigate that critical error?

Also, interestingly, I once encountered a `fatal error: concurrent map read and map write`.  [Full text (*tons* of it) here](https://gist.github.com/feuGeneA/629637df616dbe225b46318a9407157d), but it's thrown from `go-ethereum/les`.  Is that also something we should investigate?  Or perhaps just raise an Issue in go-ethereum?

In any case, here are my results:

After running the benchmark for the 400 seconds, `go tool pprof ./whisper.test ./whisper.pprof`, `web` showed that nearly half of the CPU activity is from in CGo and/or some underlying C library:
![image](https://user-images.githubusercontent.com/7883777/36026113-aeb4f0ec-0d63-11e8-97ee-67bd47b5e8cb.png)
So I ran the test through Linux `perf` to differentiate between CGo and the underlying C libary(ies):
```
gene@precision5510:~/go/src/github.com/status-im/status-go/e2e/whisper$ perf record ./whisper.test -test.bench=. -test.run=XXX -test.benchtime=400s -network=4
goos: linux
goarch: amd64
pkg: github.com/status-im/status-go/e2e/whisper
BenchmarkRequestMessagesInGroupChat-8   	     500	1011957981 ns/op
PASS
[ perf record: Woken up 1115 times to write data ]
[ perf record: Captured and wrote 279.463 MB perf.data (7325445 samples) ]
```
Note that 1011957981 ns is 1.011957981, and note that there is a 1-second sleep in the benchmark loop, so the message processing is happening quite quickly (at least, on my machine, specs below).

Examining the `perf report`, you can see that the `secp256k` library is the biggest CPU hog, with the golang `sha.keccakF1600` coming in a distant second place:
```
gene@precision5510:~/go/src/github.com/status-im/status-go/e2e/whisper$ perf report --stdio | head -39
# Total Lost Samples: 0
#
# Samples: 7M of event 'cycles:pp'
# Event count (approx.): 5087084970337
#
# Overhead  Command       Shared Object     Symbol                                                                                         
# ........  ............  ................  ...............................................................................................................................................................................................
#
    10.90%  whisper.test  whisper.test      [.] secp256k1_fe_mul_inner                                                                                                                                                                     
    10.28%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/ethereum/go-ethereum/crypto/sha3.keccakF1600                                                                                              
     6.79%  whisper.test  whisper.test      [.] secp256k1_fe_sqr_inner                                                                                                                                                                     
     5.55%  whisper.test  whisper.test      [.] secp256k1_scalar_reduce_512                                                                                                                                                                
     4.74%  whisper.test  whisper.test      [.] secp256k1_scalar_sqr_512                                                                                                                                                                   
     3.38%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/golang/snappy.encodeBlock                                                                                                                 
     3.06%  whisper.test  whisper.test      [.] secp256k1_gej_double_var                                                                                                                                                                   
     2.25%  whisper.test  whisper.test      [.] runtime.scanobject                                                                                                                                                                         
     2.22%  whisper.test  whisper.test      [.] runtime.mallocgc                                                                                                                                                                           
     1.76%  whisper.test  whisper.test      [.] runtime.memmove                                                                                                                                                                            
     1.23%  whisper.test  whisper.test      [.] runtime.heapBitsForObject                                                                                                                                                                  
     1.18%  whisper.test  whisper.test      [.] runtime.cmpbody                                                                                                                                                                            
     1.03%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/golang/snappy.decode                                                                                                                      
     0.96%  whisper.test  whisper.test      [.] runtime.nextFreeFast                                                                                                                                                                       
     0.91%  whisper.test  whisper.test      [.] runtime.heapBitsSetType                                                                                                                                                                    
     0.89%  whisper.test  whisper.test      [.] runtime.greyobject                                                                                                                                                                         
     0.88%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/syndtr/goleveldb/leveldb.(*iComparer).Compare                                                                                             
     0.85%  whisper.test  whisper.test      [.] runtime.memclrNoHeapPointers                                                                                                                                                               
     0.77%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/ethereum/go-ethereum/core/bloombits.(*Generator).AddBloom                                                                                 
     0.75%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/syndtr/goleveldb/leveldb.internalKey.ukey                                                                                                 
     0.73%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/syndtr/goleveldb/leveldb.(*iComparer).uCompare                                                                                            
     0.73%  whisper.test  whisper.test      [.] secp256k1_ge_set_xo_var                                                                                                                                                                    
     0.72%  whisper.test  whisper.test      [.] secp256k1_fe_inv                                                                                                                                                                           
     0.72%  whisper.test  whisper.test      [.] secp256k1_scalar_mul_512                                                                                                                                                                   
     0.62%  whisper.test  whisper.test      [.] secp256k1_gej_add_ge_var                                                                                                                                                                   
     0.57%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/syndtr/goleveldb/leveldb/table.(*blockWriter).append                                                                                      
     0.52%  whisper.test  whisper.test      [.] reflect.Value.Field                                                                                                                                                                        
     0.50%  whisper.test  whisper.test      [.] github.com/status-im/status-go/vendor/github.com/syndtr/goleveldb/leveldb/filter.(*bloomFilterGenerator).Generate                                                                          
```
(FYI: those percentages total to 65%. My decision to show the first 39 lines was arbitrary.)

Note that there's no mention of `cgo` in the prior listing.  Filtering the results for just mentions of `cgo`, we can see that the `cgo` overhead is negligible compared to those crypto routines:
```
gene@precision5510:~/go/src/github.com/status-im/status-go/e2e/whisper$ perf report --stdio --symbol-filter=cgo
# Total Lost Samples: 0
#
# Samples: 7M of event 'cycles:pp'
# Event count (approx.): 5087084970337
#
# Overhead  Command       Shared Object  Symbol                                                    
# ........  ............  .............  ..........................................................
#
     0.01%  whisper.test  whisper.test   [.] runtime.cgoIsGoPointer                                
     0.01%  whisper.test  whisper.test   [.] _cgo_topofstack                                       
     0.01%  whisper.test  whisper.test   [.] _cgo_1246207cbce0_Cfunc_secp256k1_ecdsa_recover_pubkey
     0.00%  whisper.test  whisper.test   [.] runtime.cgoCheckPointer                               
     0.00%  whisper.test  whisper.test   [.] runtime.cgocall                                       
     0.00%  whisper.test  whisper.test   [.] runtime.asmcgocall                                    
     0.00%  whisper.test  whisper.test   [.] runtime.cgoCheckArg                                   
     0.00%  whisper.test  whisper.test   [.] runtime.endcgo                                        
     0.00%  whisper.test  whisper.test   [.] x_cgo_sigaction                                       
```
Another thing to notice in that last listing is that there apparently is only one C library that the benchmark used, `secp256k`.

Also worth mentioning is that this test put quite the strain on my CPU, especially during the latter parts of a 10 minute run.  Subjectively, it really got the fans on my laptop revved up, whereas they're usually pretty silent.  Objectively, `top` swung between "200%" and "500%" CPU usage (what >100% means exactly, I'm not sure) on my Intel i7-6820HQ w/ 4 cores / 8 threads @ 2.7 GHz / "3.2 GHz turbo".  I wouldn't be surprised if my system was doing some CPU throttling in the name of thermal management (I've seen this system do that before).

**Comment by feuGeneA** [comment]
@JekaMas please review my previous comment and let me know if you're satisfied with this investigation or let me know what else you'd like to see.
